plugins/org.jkiss.dbeaver.ext.dameng/src/org/jkiss/dbeaver/ext/dameng/model/DamengDataSource.java

```diff
@@ -97,15 +102,88 @@ public static DBPDataKind getDataKind(@NotNull String typeName, int valueType) {
 
     }
 
+    @Nullable
+    public DamengTablespace getTablespaceById(DBRProgressMonitor monitor, long tableSpaceId) throws DBException {
+        return DamengUtils.getObjectById(monitor, tablespaceCache, this, tableSpaceId);
+    }
+
+    @Nullable
+    public DamengUser getUserById(DBRProgressMonitor monitor, long userId) throws DBException {
+        return DamengUtils.getObjectById(monitor, userCache, this, userId);
+    }
+
     @Association
     public Collection<DamengTablespace> getTablespaces(DBRProgressMonitor monitor) throws DBException {
         return tablespaceCache.getAllObjects(monitor, this);
     }
 
+    @Association
+    public Collection<DamengUser> getUsers(DBRProgressMonitor monitor) throws DBException {
+        return userCache.getAllObjects(monitor, this);
+    }
+
+    @Association
+    public Collection<DamengRole> getRoles(DBRProgressMonitor monitor) throws DBException {
+        return roleCache.getAllObjects(monitor, this);
+    }
+
     public TablespaceCache getTablespaceCache() {
         return tablespaceCache;
     }
 
+    static class RoleCache extends JDBCObjectCache<DamengDataSource, DamengRole> {
+
+        @Override
+        protected JDBCStatement prepareObjectsStatement(JDBCSession session, DamengDataSource damengDataSource) throws SQLException {
+            return session.prepareStatement("SELECT\n" +
+                    "ID,\n" +
+                    "NAME,\n" +
+                    "INFO1,\n" +
+                    "VALID,\n" +
+                    "CRTDATE\n" +
+                    "FROM\n" +
+                    "SYSOBJECTS\n" +
+                    "WHERE\n" +
+                    "TYPE$ = 'UR'\n" +
+                    "AND SUBTYPE$ = 'ROLE'\n" +
+                    "AND (INFO2 IS NULL\n" +
+                    "OR INFO2 != 1)\n" +
+                    "AND INFO1 = 0\n" +
+                    "ORDER BY\n" +
+                    "NAME");
+        }
+
+        @Override
+        protected DamengRole fetchObject(JDBCSession session, DamengDataSource damengDataSource, JDBCResultSet resultSet) throws SQLException, DBException {
+            return new DamengRole(damengDataSource, resultSet);
+        }
+    }
+
+    static class UserCache extends JDBCObjectCache<DamengDataSource, DamengUser> {
+
+        @Override
+        protected JDBCStatement prepareObjectsStatement(JDBCSession session, DamengDataSource dataSource) throws SQLException {
+            return session.prepareStatement("SELECT\n" +
+                    "USER_OBJ.*,\n" +
+                    "SYSUSERS.*,\n" +
+                    "TS.ID AS TABLESPACE_ID,\n" +
+                    "TS.NAME AS TABLESPACE\n" +
+                    "FROM\n" +
+                    "SYSOBJECTS USER_OBJ\n" +
+                    "LEFT JOIN SYSUSERS ON\n" +
+                    "SYSUSERS.ID = USER_OBJ.ID\n" +
+                    "LEFT JOIN SYS.V$TABLESPACE TS ON\n" +
+                    "USER_OBJ.INFO3 & 0x000000000000FFFF = TS.ID\n" +
```

Comment:
Could you please add a comment about this condition? It looks unusual.

Reply:
When SYSOBJECTS is a user, bit(0-1) in info3 is the default tablespace id.

Reply:
Ok. Could you please add it to the code? For the future developers :)

Reply:
Added :)

Reply:
Sorry, missed this one.